framework/src/meshgenerators/SidesetsEncloseBlocks.C

```diff
@@ -0,0 +1,148 @@
+      "It can either add sides that are not covered by a sideset by a new sidesets or"
+      "error out.");
+
+  return params;
+}
+
+SidesetsEncloseBlocks::SidesetsEncloseBlocks(const InputParameters & parameters)
+  : MeshGenerator(parameters), _input(getMesh("input"))
+{
+}
+
+std::unique_ptr<MeshBase>
+SidesetsEncloseBlocks::generate()
+{
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+
+  // Get a reference to our BoundaryInfo object for later use
+  BoundaryInfo & boundary_info = mesh->get_boundary_info();
+
+  // get a list of all sides; vector of tuples (elem, loc_side, side_set)
+  auto side_list = boundary_info.build_active_side_list();
+
+  // error on finding a side that is not covered
+  bool error_out = !isParamValid("new_boundary");
+  boundary_id_type new_sideset_id;
+  if (!error_out)
+  {
+    BoundaryName new_boundary = getParam<BoundaryName>("new_boundary");
+    std::stringstream ss;
+    ss << new_boundary;
+    ss >> new_sideset_id;
+    if (ss.fail())
+      new_sideset_id = boundary_info.get_id_by_name(new_boundary);
+
+    // make sure that _sideset exists
+    if (new_sideset_id == BoundaryInfo::invalid_id)
+      paramError("new_boundary", "Not a valid boundary");
+  }
+
+  // get blocks
+  std::vector<subdomain_id_type> vec_block_ids =
+      MooseMeshUtils::getSubdomainIDs(*mesh, getParam<std::vector<SubdomainName>>("block"));
+  std::set<subdomain_id_type> blk_ids(vec_block_ids.begin(), vec_block_ids.end());
+
+  // get boundaries
+  // check if the provided sideset name is actually a sideset id
+  // if _sideset_name can be converted to integer it's interpreted
+  // as sideset id
+  auto boundary_name_vec = getParam<std::vector<BoundaryName>>("boundary");
+  std::vector<boundary_id_type> bnd_ids_vec;
+  bnd_ids_vec.reserve(boundary_name_vec.size());
+  std::set<boundary_id_type> bnd_ids;
+  for (auto & bnd_name : boundary_name_vec)
+  {
+    std::stringstream ss;
+    boundary_id_type sideset_id;
+    ss << bnd_name;
+    ss >> sideset_id;
+    if (ss.fail())
+      sideset_id = boundary_info.get_id_by_name(bnd_name);
+
+    // make sure that _sideset exists
+    if (sideset_id == BoundaryInfo::invalid_id)
+      paramError("boundary", "Not a valid boundary");
+    bnd_ids_vec.push_back(sideset_id);
+    bnd_ids.insert(sideset_id);
+  }
+
+  // loop over all elements in blocks and for each elem over each side
+  // and check the neighbors
+  for (auto & block_id : blk_ids)
+  {
+    for (const Elem * elem : as_range(mesh->active_local_subdomain_elements_begin(block_id),
+                                      mesh->active_local_subdomain_elements_end(block_id)))
+    {
+      // loop through sides
+      for (unsigned int j = 0; j < elem->n_sides(); ++j)
+      {
+        const Elem * neigh = elem->neighbor_ptr(j);
+
+        // is this an outside boundary to blocks?
+        // NOTE: the next line is NOT sufficient for AMR!!
```

Comment:
can I have some help from maybe @lindsayad. I don't think the check I am doing will work if we have hanging nodes. How do I generalize this correctly?

Reply:
Are you trying to support the case where you have subdomains that differ on different children of the same parent element?  libMesh supports that case, but unfortunately since we don't yet support sidesets that aren't on sides of parent elements, you won't yet be able to support that case here.  For now it probably makes sense to just worry about level-0.

I'd worry more about distributed mesh support, personally.  For the sides in between ghost elements and remote elements on any particular processor's part of the mesh, you don't have enough information to tell whether you need a new boundary id there or not without communication.  The solutions to this are "exit with an error if the mesh is distributed", "throw in a MeshSerializer to temporarily un-distribute it" or "write that communication routine", ranked from least to most by both difficulty and usefulness.

Reply:
My concern was this case

![Screen Shot 2022-03-30 at 12 29 43 PM](https://user-images.githubusercontent.com/7524371/160906085-a0680f5c-f3f5-41e7-982e-1e2e84efceff.png)

Both elements are in the same block. But between elem 2 is refined. The side is checked
with this line of code:
```
bool is_outer_bnd = !neigh || blk_ids.find(neigh->subdomain_id()) == blk_ids.end();
```
My fear was that somehow `neigh == nullptr` in that case. I now think that `neigh` is not
`nullptr` but that it is simply not an active element. Is that correct @roystgnr?

I think that is different than what you thought I 
wanted. Is that the case you described:
![Screen Shot 2022-03-30 at 12 35 23 PM](https://user-images.githubusercontent.com/7524371/160906957-296a064a-fea0-49d5-b34e-8ceac5d7f0c8.png)
i.e., a single parent element has children with different blocks. That is too exotic of a case for me to care about.

Reply:
Actually if I think about it. What subdomain is returned for this case:
![Screen Shot 2022-03-30 at 12 38 08 PM](https://user-images.githubusercontent.com/7524371/160907411-076459eb-186d-48ab-87c7-aaa7a1bb8e56.png)
I am in elem 1 and for the indicated side I get the neighbor pointer. Then I get the subdomain id. What is that subdomain id? (B1, B2, B3)?

Reply:
> I'd worry more about distributed mesh support, personally. For the sides in between ghost elements and remote elements on any particular processor's part of the mesh, you don't have enough information to tell whether you need a new boundary id there or not without communication. The solutions to this are "exit with an error if the mesh is distributed", "throw in a MeshSerializer to temporarily un-distribute it" or "write that communication routine", ranked from least to most by both difficulty and usefulness.

I am not sure I understand why that is. I don't think I ever access sides of ghosted elements because my loops are only over active local elems.

Reply:
In the first case, where I'll assume you got four children by refining Elem 2 and you haven't edited any of their subdomain ids since, I think you're safe: Elem 2 still has the same subdomain as all it's children, and Elem 2 is still the neighbor of Elem 1, it's just that `elem2->active()` would now be false for it.

In the second case?  You're not at all safe: for all you know, you might get **B4** for the subdomain id!  `Elem::subdomain_id()` just returns the subdomain `this` element, regardless of whether it's active or not, and regardless of what ids its descendants might have.  If, for instance, elem2 is on block 4, and you refine it and set its children to live on blocks 1,2, and 3, then elem2 is *still* on block 4!  You can get its set of active descendants (or its set of active descendants on the side neighboring elem1; that's a natural enough task that we have an API for it) and check each of their `subdomain_id()` values individually, but we won't do anything to try to keep the parent in sync here; in cases like this where you give different subdomains to different children it's impossible to keep the parent in sync with all of them.